Update start servers

saijel opened the pull request:
list of changes:
- remove the dependency of gnome-terminal in services makefiles
- add the ability to start all services (secure and non-secure hosts) using a tmux session with a tmux window for each service
- add the ability to make the host secure in order to add the ASR functionality by generating self-signed certificates for the host

jhauswald commented on lucida/commandcenter/app.py:
```diff
@@ -1,17 +1,18 @@
 import logging
 
 
+
```
🤔

saijel replied:
I'll remove all of these extra white spaces. I stupidly diff -w to ignore the whitespace.

jhauswald commented on lucida/commandcenter/app.py:
```diff
@@ -27,20 +28,40 @@
+                threaded=True, ssl_context=('certs/server.crt', 'certs/server.key'))
+    # For http
+    else:
+        print 'Starting non-secure flask'
```
So non-secure means you can't use ASR right? Is that stated somewhere?

saijel replied:
I'm adding a comment within this file but also I think it'll be more clear once I write out the instructions on how to start servers with and without ASR.

jhauswald commented on lucida/calendar/Makefile:
```diff
@@ -24,7 +24,7 @@ corenlp:
 
 start_server: 
-	gnome-terminal -x bash -c "./gradlew run $(filter-out $@,$(MAKECMDGOALS)); read -n1"
+	./gradlew run
```
can you speak to what `$(filter-out $@,$(MAKECMDGOALS)); read -n1` was doing?

saijel replied:
I honestly don't know what these were doing. Every time I run with  `$(filter-out $@,$(MAKECMDGOALS)); read -n1` it says these commands are not found. 

According to the internet: `$(filter-out PATTERN...,TEXT)` returns all whitespace-separated words in `TEXT` that do not match any of the `PATTERN` words, removing the words that do match one or more. And `read -n1` means read one character at a time. But `MAKECMDGOALS` is not defined anywhere.

jhauswald commented on lucida/speechrecognition/kaldi_gstreamer_asr/simple_start.sh:
```diff
@@ -1,4 +1,6 @@
 
-gnome-terminal -x bash -c "python kaldigstserver/worker.py -u ws://localhost:8081/worker/ws/speech -c sample_english_nnet2.yaml; read -n1"
+python kaldigstserver/worker.py -u ${ASR_ADDR_PORT}/worker/ws/speech -c sample_english_nnet2.yaml
+
```
why these extra lines?

saijel replied:
Getting rid of the extra spaces.

jhauswald commented on tools/start_all_tmux.sh:
```diff
@@ -0,0 +1,71 @@
+if [ "$1" == "secure" ]; then
+    echo "Enabling secure host"
+    # Getting the host IP address
+    export ASR_ADDR_PORT="wss://$(/sbin/ifconfig eth0 | grep 'inet addr:' | cut -d: -f2 | awk '{ print $1}'):8081"
```
This makes me a little nervous, so all ASR ports run on 8081 and its sort of hard coded all over the place right?

saijel replied:
Yes, that's how it seems. Each service has a port that is hard-coded and used in multiple places.

jhauswald replied:
Yes I know that I'm asking if you have an idea for a better design?

saijel replied:
The issue is we need a common place to have a service registry that all servers and cmd center can access. In some other implementations the command center holds the registry, the microservices register themselves giving the cmd their host and port, and the cmd adds this to its service registry. This way the microservices can change their locations and just register themselves every time they are initialized.

jhauswald commented on tools/start_all_tmux.sh:
```diff
@@ -0,0 +1,71 @@
+    unset SECURE_HOST
+fi
+
+declare -a commandcenter=("CMD" "$(pwd)/../lucida/commandcenter/")
```
I don't know what declare does, can you enlighten me?

saijel replied:
the declare allows you to declare and initialize an array by specifying the list of elements (separated by white space) rather than initializing each element of an array separately

jhauswald commented on lucida/djinntonic/dig/Makefile:
```diff
@@ -23,7 +23,7 @@ $(TARGET): $(OBJECTS)
+	./DIGServer
 
 start_test: 
 	gnome-terminal -x bash -c "./run-test.sh $(filter-out $@,$(MAKECMDGOALS)); read -n1"
```
if you're removing all gnome-terminals, why not get rid of this one too?

saijel replied:
There currently isn't a automated way to test all of the servers so when someone wants to run tests they have to go to each individual microservice folder and run start_test. In this situation it may be necessary to have a new terminal open as it won't be necessary to keep it running after the test has been run unlike a server.

jhauswald replied:
I see your point but I think in that case there should be a better way to test. can you open an issue about moving `start_test` as unittest, similar to what we do internally? Look at the infrastructure we use

saijel replied:
Okay, I created an issue. In the meantime let me look into our internal infrastructure.

jhauswald commented on Makefile:
```diff
@@ -19,3 +19,9 @@ local:
 
 start_all:
```
are these other make targets still relevant? In adding your features, are there any cleanups you can make that would make it easier to understand what you're adding?

jhauswald replied:
Think about that across your PR. I know the make targets are defined one way, unclear whether its the best way

jhauswald replied:
You see this comment?

saijel replied:
I will remove `start_all` since `start_all_tmux` replaces that. As far as the way make targets are defined, they are defined in a consistent way across all services which makes it easy to loop through all services and start them.

jhauswald replied:
sounds good

saijel replied:
Once the PR is merged I will be sure to update the wikis/README to show the new way to start all servers (both with and without ASR).

jhauswald commented on tools/start_all_tmux.sh:
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
```
💯

jhauswald commented on tools/start_all_tmux.sh:
```diff
@@ -0,0 +1,71 @@
+    facerecognition)
+
+# Create the session
+tmux new-session -s ${SESSION_NAME} -n vim -d
```
Tmux is an okay way of doing this, have you thought/investigated other ways of deploying/running web services? I know about [supervisord](http://supervisord.org/). What else is out there?

jhauswald replied:
@saijel you get a chance to think about this?

saijel replied:
I've gotten as far seeing different options out there. This article has a good comparison of different options (http://centos-vn.blogspot.com/2014/06/daemon-showdown-upstart-vs-runit-vs.html). I'll need to do more research about what would be best for our situation.

jhauswald commented:
I still need to test all this out but there are some discussion points as well

jhauswald commented:
After testing:
- what's the first vim window for?
- can you add a check if I already have a window named lucida?
- can you add a check to not nest tmux sessions if i'm already in tmux? This is generally bad practice.
- how do i got about adding a new service? is that automatic or do I need to edit the bash script everytime? Is this related to your registry idea?

saijel commented:
Answers below.

---
Saijel Mokashi
https://clinc.com

On Mon, Apr 17, 2017 at 8:29 AM, Johann Hauswald <notifications@github.com>
wrote:

> After testing:
>
>    - what's the first vim window for? *I added this window to do
>    development in. I guess it is not completely necessary. Should I remove it?*
>    - can you add a check if I already have a window named lucida? *Yes, I
>    added this check and updated the pull request.*
>    - can you add a check to not nest tmux sessions if i'm already in
>    tmux? This is generally bad practice. *Fixed in the above statement*
>    - how do i got about adding a new service? is that automatic or do I
>    need to edit the bash script everytime? Is this related to your registry
>    idea?* Currently, you'd have to edit the bash script every time. I
>    think the entire process of adding a microservice needs to be planned out
>    to try to make it more automated. *
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/claritylab/lucida/pull/163#issuecomment-294475331>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AZQku_nKUCzHuCW5sYZF0_cuWpvT4R4fks5rw1s2gaJpZM4MztZ6>
> .
>

saijel commented:
@jhauswald

jhauswald commented:
- what's the first vim window for? *I added this window to do
    development in. I guess it is not completely necessary. Should I remove it?*
yes
    - can you add a check if I already have a window named lucida? *Yes, I
    added this check and updated the pull request.*
nice
    - can you add a check to not nest tmux sessions if i'm already in
    tmux? This is generally bad practice. *Fixed in the above statement*
nice
    - how do i got about adding a new service? is that automatic or do I
    need to edit the bash script everytime? Is this related to your registry
    idea?* Currently, you'd have to edit the bash script every time. I
    think the entire process of adding a microservice needs to be planned out
    to try to make it more automated. *
let's discuss further

jhauswald commented on tools/start_all_tmux.sh:
```diff
@@ -0,0 +1,80 @@
+# use the following command:
+# $tmux a -t lucida
+
+SESSION_NAME="lucida"
```
where's the check to see if you're already in a tmux session? What if I'm in a tmux session that is not called lucida?

jhauswald commented:
Also what about the fixes that you have? can you make a quick MR for those?

jhauswald commented on tools/start_all_tmux.sh:
```diff
@@ -0,0 +1,82 @@
+    facerecognition)
+
+# First window (0) -- vim
+tmux send-keys -t ${SESSION_NAME} 'vim' C-m
```
are you still generating a vim window? I don't think there's much use...

jhauswald commented on tools/start_all_tmux.sh:
```diff
@@ -0,0 +1,82 @@
+    echo "Session ${SESSION_NAME} already exists."
+    exit 0;
+elif [ -n "$TMUX" ]; then
+    echo "Already in a tmux session"
```
💯
